src/main/java/com/wavefront/sdk/direct/ingestion/WavefrontDirectIngestionClient.java

```diff
@@ -175,8 +193,55 @@ public void flush() throws IOException {
 
   private void internalFlush(LinkedBlockingQueue<String> buffer, String format)
       throws IOException {
+    for (List<String> batch : getBatches(buffer)) {
+      internalFlushBatch(buffer, format, batch);
+    }
+  }
 
-    List<String> batch = getBatch(buffer);
+  private List<List<String>> getBatches(LinkedBlockingQueue<String> buffer) {
+    List<String> block = getBlock(buffer);
+
+    if (messageSizeBytes == null) {
+      return Collections.singletonList(block);
+    }
+
+    List<List<String>> batches = new ArrayList<>();
+    List<String> batch = new ArrayList<>();
+    int numBytesInBatch = 0;
+
+    for (String item : block) {
+      int numBytes = item.getBytes().length;
+      if (numBytesInBatch + numBytes > messageSizeBytes) {
+        if (!batch.isEmpty()) {
+          batches.add(batch);
+        }
+        batch = new ArrayList<>();
+        numBytesInBatch = 0;
+      }
+      if (numBytes > messageSizeBytes) {
+        logger.log(Level.WARNING,
+                "Unable to report data that exceeds messageSizeKilobytes: " + item);
+      } else {
+        batch.add(item);
+      }
+      numBytesInBatch += numBytes;
+    }
+    if (!batch.isEmpty()) {
+      batches.add(batch);
+    }
+
+    return batches;
+  }
+
+  private List<String> getBlock(LinkedBlockingQueue<String> buffer) {
+    int blockSize = Math.min(buffer.size(), batchSize);
+    List<String> points = new ArrayList<>(blockSize);
+    buffer.drainTo(points, blockSize);
```

Comment:
Why do we have to move data over to another location for processing? Why can't we use the `buffer` object directly? Because if someone set `batchSize` to `Integer.MAX` and when it is nearly full and the `getBlock` operation is called, we are forcing it to simply copy every element into another place.

Reply:
Consolidated the processing in `getBatch`.